server: support escaping separator in search terms.

Ionic opened the pull request:
For non-regex search terms, the separator is hardcoded to a comma character (which is fine), but there has been no way to escape it until now. Since commas are legal characters in references (and probably other fields), having a way to search for them can actually be useful.

While one could argue that regular expressions can be used to achieve that, typical users have no idea how that works and often get them wrong, so providing an easy alternative sounds beneficial.

willuhn commented:
Danke für das Patch. Du hättest das aber nicht auf Englisch machen müssen. Hibiscus wird ja ohnehin nur in Deutschland verwendet ;)

PS: Ich glaube, mit dem Iterator hast du es dir unnötig schwer gemacht. Ich denke, mit einer einfachen for-Schleife in etwa der Form kannst du ja einfach auf das voherige oder nächste Element schauen:

```
for (int i=0;i<split.size();++i)
{
  String elem = split.get(i);
  boolean new_merge = split.size() > i && elem.endsWith("\\");
      if (new_merge) {
        elem = StringUtils.chop(elem);
      }  
}
```

Dann sparst du dir das previous/next/hasNext, um die internen Pointer in den Iteratoren zu positionieren.

Ionic commented:
Ich finde, dass Software möglichst zugänglich geschrieben sein sollte und Deutsch erfüllt das für mich nicht. Gerade, weil durch SEPA Hibiscus die Chance hat auch außerhalb Deutschlands Fuß zu fassen, wenn man sich eh schon auf einen europäischen Standard geeinigt hat. Ich denke auch bei der Programmierung nicht auf Deutsch - falls man das so beschreiben kann.

Jein. Die split-Liste ist ja gar kein Problem mit den Java-Iteratoren, nur die ret-Liste, die ich aufbaue. Das hab ich zwar initial auch mit einfachen Indices gemacht (auf -1 initialisiert und dann inkrementiert, wenn etwas hinzugefügt werden sollte), aber irgendwas in mir schreit immer, dass man indexbasierte Schleifen gar nicht erst mehr schreiben sollte. Ich habe seit 10 Jahren nichts mehr mit Java gemacht und Iteratoren in C++ sind sehr viel schöner modelliert, aber so geht es ja auch.

Was mich eher interessiert, ob du die Funktionalität überhaupt haben möchtest. Ich hab irgendwie damit gerechnet, dass du sagst, dass das gar nicht gebraucht wird, weil man dieselbe Funktionalität ja mit dem Regex-Feature bekommt, aber "normale Nutzer" haben damit große Probleme und meistens sind Regex ja auch Overkill.

willuhn commented:
> Ich finde, dass Software möglichst zugänglich geschrieben sein sollte und Deutsch erfüllt das für mich nicht. Gerade, weil durch SEPA Hibiscus die Chance hat auch außerhalb Deutschlands Fuß zu fassen, wenn man sich eh schon auf einen europäischen Standard geeinigt hat. Ich denke auch bei der Programmierung nicht auf Deutsch - falls man das so beschreiben kann.

SEPA wird ja EU-weit verwendet. Das Transport-Protokoll FinTS ist aber nach wie vor eine rein deutsche Erfindung. Und in all den Jahren hat es nie den Weg über die Grenze geschafft. 
 
> Jein. Die split-Liste ist ja gar kein Problem mit den Java-Iteratoren, nur die ret-Liste, die ich aufbaue. Das hab ich zwar initial auch mit einfachen Indices gemacht (auf -1 initialisiert und dann inkrementiert, wenn etwas hinzugefügt werden sollte), aber irgendwas in mir schreit immer, dass man indexbasierte Schleifen gar nicht erst mehr schreiben sollte. Ich habe seit 10 Jahren nichts mehr mit Java gemacht und Iteratoren in C++ sind sehr viel schöner modelliert, aber so geht es ja auch.

Die Iteratoren sind in Java eine veraltete Karteileiche. Zwischenzeitlich waren Implementierungen des Collection-Interfaces das Mittel der Wahl. Inzwischen ist es java.util.stream. Es spricht aber auch heute noch nichts gegen eine einfache for-Schleife mit Index oder ein "for (e in list)".
 
> Was mich eher interessiert, ob du die Funktionalität überhaupt haben möchtest. Ich hab irgendwie damit gerechnet, dass du sagst, dass das gar nicht gebraucht wird, weil man dieselbe Funktionalität ja mit dem Regex-Feature bekommt, aber "normale Nutzer" haben damit große Probleme und meistens sind Regex ja auch Overkill.

Dein Feature ist ja durchaus sinnvoll und ich übernehme es auch gern. Ich denke aber, dass es mit weniger Code umsetzbar ist. String#split kann direkt ein Regex übergeben werden. Wenn es also nur darum geht, den String an Kommas zu trennen - hierbei aber all jene Kommas ignoriert werden sollen, die mit einem "\" escaped sind, dann würde in Zeile 327 statt:

```
String[] list = s.toLowerCase().split(","); // Wir beachten Gross-Kleinschreibung grundsaetzlich nicht
```

schon etwas in der Art reichen:

```
String[] list = s.toLowerCase().split("[^\\\\],"); // Wir beachten Gross-Kleinschreibung grundsaetzlich nicht
```
Die Methode "split_escape" wäre dann gar nicht nötig. Es sei denn, ich hab noch irgendwas übersehen.

Ionic commented:
> Dein Feature ist ja durchaus sinnvoll und ich übernehme es auch gern. Ich denke aber, dass es mit weniger Code umsetzbar ist. String#split kann direkt ein Regex übergeben werden. Wenn es also nur darum geht, den String an Kommas zu trennen - hierbei aber all jene Kommas ignoriert werden sollen, die mit einem "" escaped sind, dann würde in Zeile 327 statt:
> 
> ```
> String[] list = s.toLowerCase().split(","); // Wir beachten Gross-Kleinschreibung grundsaetzlich nicht
> ```
> 
> schon etwas in der Art reichen:
> 
> ```
> String[] list = s.toLowerCase().split("[^\\\\],"); // Wir beachten Gross-Kleinschreibung grundsaetzlich nicht
> ```

Wieso `\\\\`? Eigentlich doch `\\`, weil man nur einen echten Backslash braucht und nicht zwei.


> Die Methode "split_escape" wäre dann gar nicht nötig. Es sei denn, ich hab noch irgendwas übersehen.

Ja, ich weiß. Dachte ich zuerst auch, bis mir dann eingefallen ist, dass wir die Escape-Chracters ja noch loswerden wollen (aber auch nur, wenn sie speziell sind) und es so nicht reicht.

Dies wäre an sich kein großes Problem, da ein weiteres `.replaceAll("\\" + separator, separator);` dafür ausreichen sollte. Sofern ich keinen Sonderfall DABEI übersehen habe...

Allerdings gibt es beim Splitting dann doch noch einige Sonderfälle zu beachten, z. B., dass am Anfang evtl. gar nichts steht, also Input wie `",bla\,...,blubb"`, was dann als erstes Element `",bla,..."` zurückliefern würde, was falsch wäre. Man müsste also eher sowas wie `"(^|[^\\])" + separator` verwenden. Der nächste Sonderfall wären dann wieder leere Elemente, wie als Input-String `",,,"`, die mit dem Regex auch nicht korrekt funktionieren würden. Also müsste es sowas wie `"(^|[^\\]?)" + separator` sein, aber ich hatte Angst, dass das wieder effektiv einfach nur `,` bedeutet, ohne es vorher groß getestet zu haben.

Vermutlich hätte man auch einen Regex finden können, der das Splitting in allen Fällen fehlerfrei vornimmt, aber das war mir ehrlich gesagt zu fehleranfällig und zu viel eingebettete Magie, also, dass man genau wissen muss, wieso das alles da drin wichtig ist, ohne es auf den ersten Blick sehen zu können.

Nach all diesen Überlegungen hab ich mir gedacht, dass es besser wäre normal aufzuteilen und dann die Einträge wieder zusammenzuführen, da die Regeln zur Zusammenführung sehr viel einfacher sind als das Aufteilen mit Escaping per Regex. Das ist zwar mehr Code, aber viel klarer in der Funktionsweise und man schießt sich nicht so einfach ins Knie.

Was mein Code jetzt noch anders macht als die Vorgängerversion ist leere Einträge zu behalten. Ich ändere das mal noch schnell. Das originale Splitting MUSS leere Elemente beibehalten, damit das Merging richtig funktioniert (Sonderfall am Ende, vermutlich gibt es noch andere Gründe, die mir entfallen sind), aber an anderen Stellen kann ich die leeren Elemente wieder rausfiltern.

willuhn commented:
Hab's übernommen. Die doppelten "\" sind in Java nötig, weil der Backslash doppelt escaped werden muss. Einmal für den Regex (um diesem zu sagen, dass wirklich der Backslash gemeint ist) und dann nochmal für das String-Literal von Java selbst.

Ionic commented:
Danke! Oh, ja, double quotes... ich mach zu viel in der Shell mit single quotes...
